### PR TITLE
extensions: add OpenTelemetry tracer extension for exporting OTLP traces

### PR DESCRIPTION
**Pull request by AlexanderEllis**
Commit Message: Add OpenTelemetry Tracer Extension for exporting OTLP traces
Additional Description: This adds a new OpenTelemetry Tracer that exports OTLP traces via Envoy's async gRPC mechanism. It includes `traceparent` header parsing, buffering, and basic stats.
Risk Level: Medium (new extension that is not enabled)
Testing: Added unit tests for the new tracer and performed manual testing with the OpenTelemetry Collector
Docs Changes: TODO
Release Notes: opentelemetry: add OpenTelemetry tracer extension
Platform Specific Features: N/A
Part of [#9958](https://github.com/envoyproxy/envoy/issues/9958)

Possible fix for #20641

### CODE REVIEW

File: source/extensions/tracers/opentelemetry/tracer.cc
Comment (phlax) [inline source/extensions/tracers/opentelemetry/tracer.cc:138]: missing newline - the easiest is to fix your editor to prevent this

**Comment by moderation** [comment]
Fantastic to see this! As part of this change can you update to the latest version of OpenTelemetry. We are currently on version `0.11.0` - the latest is [`0.14.0`](https://github.com/open-telemetry/opentelemetry-proto/releases/tag/v0.14.0)

https://github.com/envoyproxy/envoy/blob/main/api/bazel/repository_locations.bzl#L99-L109

File: source/extensions/tracers/opentelemetry/tracer.cc
Comment (AlexanderEllis) [reply source/extensions/tracers/opentelemetry/tracer.cc]: Ah whoops yes of course, thanks for catching that! Must have missed this in my editor settings.

File: api/bazel/repository_locations.bzl
Comment (moderation) [inline api/bazel/repository_locations.bzl:105]: `0.15.0` has just been released - https://github.com/open-telemetry/opentelemetry-proto/releases/tag/v0.15.0

**Review by moderation — CHANGES_REQUESTED** [review]

File: api/envoy/config/trace/v3/opentelemetry.proto
Comment (lizan) [inline api/envoy/config/trace/v3/opentelemetry.proto:24]: Can you add comment for this field? Also it should be required?

**Review by lizan — COMMENTED** [review]
api review
/wait

**Comment by K-Prabha** [comment]
Hi - @AlexanderEllis - any update on this?

File: api/bazel/repository_locations.bzl
Comment (AlexanderEllis) [reply api/bazel/repository_locations.bzl]: Updated this to 0.15.0 👍 

I noticed that the bump to 0.15.0 deprecated a field used in this code (and in the access_logger), `InstrumentationLibrary{Logs,Spans}` in favor of `Scope{Logs,Spans}`, and it's planned to remove the former in June. Should I update those here as well, or should that be a part of the version update when those fields are actually removed? 

Wasn't sure if that would be too many extra changes here (and I'd also be happy to do so in a follow up).

File: api/envoy/config/trace/v3/opentelemetry.proto
Comment (AlexanderEllis) [reply api/envoy/config/trace/v3/opentelemetry.proto]: Ah great call, done and done.

**Comment by pseudonator** [comment]
@AlexanderEllis How are you dealing with attributes per span ?

File: api/bazel/repository_locations.bzl
Comment (htuch) [reply api/bazel/repository_locations.bzl]: Suggest splitting this into a separate PR.

**Review by moderation — DISMISSED** [review]
/lgtm deps

**Comment by phlax** [comment]
@htuch reassigned to you for final dep-shepherd approval

**Review by htuch — COMMENTED** [review]
/lgtm deps

**Review by htuch — COMMENTED** [review]
/lgtm api

**Comment by K-Prabha** [comment]
Is this ready for approve/merge?

**Comment by moderation** [comment]
Looking forward to this one. /cc @yanavlasov, @lizan, @phlax

File: source/extensions/tracers/opentelemetry/config.cc
Comment (lizan) [inline source/extensions/tracers/opentelemetry/config.cc:22]: Where is this limitation comes from? Also this need to be documented in proto file.

File: api/envoy/config/trace/v3/opentelemetry.proto
Comment (lizan) [inline api/envoy/config/trace/v3/opentelemetry.proto:25]: I don't see the reason that only Google gRPC is supported, why is that?

File: api/bazel/repository_locations.bzl
Comment (phlax) [inline api/bazel/repository_locations.bzl:103]: version `0.16.0` is now released

**Comment by phlax** [comment]
@AlexanderEllis this needs main merge

File: source/extensions/tracers/opentelemetry/config.cc
Comment (AlexanderEllis) [reply source/extensions/tracers/opentelemetry/config.cc]: Ah on further review, I think I was fooling myself by following the pattern in the OpenCensus tracer. I don't believe this limitation applies here, and I've removed it. Thanks!

File: api/envoy/config/trace/v3/opentelemetry.proto
Comment (AlexanderEllis) [reply api/envoy/config/trace/v3/opentelemetry.proto]: I think I copied this from the other comment as well, but after taking another look, I also don't think this applies. Removed this warning.

File: api/bazel/repository_locations.bzl
Comment (AlexanderEllis) [reply api/bazel/repository_locations.bzl]: I'm getting some good practice at updating the OTLP proto version :) Updated, thanks!

File: source/extensions/tracers/opentelemetry/tracer.cc
Comment (lizan) [inline source/extensions/tracers/opentelemetry/tracer.cc:21]: This is a source of test flakiness because of the static initialization fiasco. Need to be `constexpr absl::string_view`.

Comment (AlexanderEllis) [reply source/extensions/tracers/opentelemetry/tracer.cc]: Nice, done.

**Comment by moderation** [comment]
@AlexanderEllis Looks like CI is failing on the new zstd example

`2022-04-15T17:16:53.0555962Z [36menvoy-stats_1  |[0m Unable to parse JSON as proto (INVALID_ARGUMENT:(compressor_library.typed_config): invalid value Invalid type URL, unknown type: envoy.extensions.compression.zstd.compressor.v3.Zstd for type Any):`

May need a main merge?

**Comment by jmarantz** [comment]
as maintainer on-call I'm trying to follow up on stalled PRs. I think this one got senior maintainer approval but only for API & dependencies, and has not had a deep review.

Assigning to @phlax for now but lmk if that was the wrong assignee.

Also CI is not totally happy yet.

**Comment by phlax** [comment]
@AlexanderEllis it looks to me like the CI fail is a bad merge - could you merge main please ?

**Comment by AlexanderEllis** [comment]
@phlax Thanks for the reminder! Merged main and it looks like the CI is happy (had to rerun what looked like a TSAN flake, but it's all passing).

**Review by moderation — APPROVED** [review]
Re-approved deps. Needs an API approval now. Will be great to land this and hope we can merge soon. Thanks @AlexanderEllis

File: api/envoy/config/trace/v3/opentelemetry.proto
Comment (htuch) [inline api/envoy/config/trace/v3/opentelemetry.proto:16]: Nit: you don't need this.

**Review by htuch — COMMENTED** [review]
/lgtm api (will need to reapprove when nit is fixed)

File: api/envoy/config/trace/v3/opentelemetry.proto
Comment (AlexanderEllis) [reply api/envoy/config/trace/v3/opentelemetry.proto]: Nice, thanks! Removed.

**Review by htuch — COMMENTED** [review]
/lgtm api

**Comment by ethernoy** [comment]
Hi, I wonder if this PR adds OpenTelemetry trace instrumentation on the gRPC call from envoy to remote rate limit service?

**Comment by moderation** [comment]
`opentelemetry_proto` `0.17.0` is out https://github.com/envoyproxy/envoy/issues/21190

**Comment by K-Prabha** [comment]
@phlax - I have not followed all details in the thread. Is there a way we can iteratively progress on this instead of pausing it. We are one of the adopters of envoy and we are really looking for OTel support in Envoy

**Comment by htuch** [comment]
I had assumed @lizan was owning the review. I can take a pass today as well and merge if possible.

**Comment by austinlparker** [comment]
Heya, quick question -- https://github.com/envoyproxy/envoy/pull/21336#issuecomment-1129112184 is discussing removing the (now deprecated) OpenTracing-based Lightstep tracer from Envoy. Would this PR land in the same release as that deprecation?

**Comment by jmacd** [comment]
The issue raised by @austinlparker is that https://github.com/envoyproxy/envoy/pull/21370 would delete a tracer that is still in use before this tracer replaces it for our customers. Can we be sure this merges in the same release as #21370?

File: source/extensions/tracers/opentelemetry/span_context_extractor.cc
Comment (lizan) [inline source/extensions/tracers/opentelemetry/span_context_extractor.cc:54]: In general, no exceptions in data path?

File: source/extensions/tracers/opentelemetry/span_context_extractor.cc
Comment (lizan) [inline source/extensions/tracers/opentelemetry/span_context_extractor.cc:61]:
```suggestion
  std::vector<absl::string_view> propagation_header_components =
```

This function copies those string multiple times. You may want to make SpanContext takes string_view as constructor parameter.

File: api/bazel/repository_locations.bzl
Comment (moderation) [inline api/bazel/repository_locations.bzl:105]: `0.18.0` is now out - https://github.com/envoyproxy/envoy/issues/21351

**Review by moderation — CHANGES_REQUESTED** [review]
Update to latest `opentelemetry_proto`

File: source/extensions/tracers/opentelemetry/grpc_trace_exporter.h
Comment (yanavlasov) [inline source/extensions/tracers/opentelemetry/grpc_trace_exporter.h:69]: It would be good to document in the proto config that tracer drops traces if server does not read data fast enough.

File: source/extensions/tracers/opentelemetry/span_context_extractor.cc
Comment (yanavlasov) [reply source/extensions/tracers/opentelemetry/span_context_extractor.cc:54]: +1, you can use the `absl::StatusOr` for this case.

File: source/extensions/tracers/opentelemetry/span_context_extractor.cc
Comment (yanavlasov) [inline source/extensions/tracers/opentelemetry/span_context_extractor.cc:66]:
```suggestion
  absl::string_view version = propagation_header_components[0];
```

**Review by yanavlasov — CHANGES_REQUESTED** [review]
/wait

File: api/bazel/repository_locations.bzl
Comment (AlexanderEllis) [reply api/bazel/repository_locations.bzl]: Updated, thanks!

File: source/extensions/tracers/opentelemetry/grpc_trace_exporter.h
Comment (AlexanderEllis) [reply source/extensions/tracers/opentelemetry/grpc_trace_exporter.h:69]: Done.

File: source/extensions/tracers/opentelemetry/span_context_extractor.cc
Comment (AlexanderEllis) [reply source/extensions/tracers/opentelemetry/span_context_extractor.cc]: Updated.

File: source/extensions/tracers/opentelemetry/span_context_extractor.cc
Comment (AlexanderEllis) [reply source/extensions/tracers/opentelemetry/span_context_extractor.cc]: Ah great call, updated this and in the usage/tests. Thanks!

**Comment by AlexanderEllis** [comment]
Sorry for the delay here, ended up very busy and had to reconfigure my dev setup on a new computer.

@jmacd @austinlparker  My only concern with the addition/removal swap is that this new tracer will be initially have the `wip` status (pending a follow up to handle `tracestate`) and would then be in `alpha` without substantial burn time (as per https://github.com/envoyproxy/envoy/blob/main/EXTENSION_POLICY.md#extension-stability-and-security-posture), while the lightstep tracer is already `stable`. I'm not familiar with the PR/release timing, but I did just wanted to flag that for you.

**Comment by yanavlasov** [comment]
@AlexanderEllis sanitizer failures look legit: https://source.cloud.google.com/results/invocations/53ad609c-54e2-4184-ba66-3ad91d4714e7/targets/%2F%2Ftest%2Fextensions%2Ftracers%2Fopentelemetry:opentelemetry_test/log

**Comment by yanavlasov** [comment]
@AlexanderEllis what are your plans for finishing this? If you do not have time, would it be ok if someone else can complete the PR as it is blocking protobuf upgrade?

**Comment by AlexanderEllis** [comment]
@yanavlasov Thanks for the heads up! Sorry for the delay, didn't realize this was blocking anything other than itself. I should have some time to dig into the failures this weekend. After that, I'll be afk for ~2 weeks, and hopefully things should be more stable after that. That being said, if that doesn't work timing wise, totally fine if anyone else jumps in on the PR — it's all Envoy's code anyways :)

File: source/extensions/tracers/opentelemetry/span_context.h
Comment (yanavlasov) [inline source/extensions/tracers/opentelemetry/span_context.h:34]: I think you can return `const std::string&` to avoid copying. Other accessors as well.

**Review by yanavlasov — CHANGES_REQUESTED** [review]
LGTM modulo small not.

File: source/extensions/tracers/opentelemetry/span_context.h
Comment (AlexanderEllis) [reply source/extensions/tracers/opentelemetry/span_context.h]: Nice, good call! Updated.

**Review by phlax — APPROVED** [review]
lgtm, thanks so much @AlexanderEllis for staying the journey

landing cos i was asked, and im drunk enuf 8/

cc @moderation

**Comment by K-Prabha** [comment]
🎉 Thank you @AlexanderEllis for the persistence and thank you community for helping this push thru. Much appreciated

**Comment by phlax** [comment]
eek @AlexanderEllis looks like i broke `main` landing this - i reckon minor hiccup - any ideas on how we can fix forward ?

```console

unknown file: Failure
C++ exception with description "Protobuf message (type opentelemetry.proto.collector.logs.v1.ExportLogsServiceRequest reason INVALID_ARGUMENT:(resource_logs.instrumentation_library_logs[0]) logs: Cannot find field.) has unknown fields" thrown in the test body.
test/extensions/access_loggers/open_telemetry/grpc_access_log_impl_test.cc:51: Failure
Actual function call count doesn't match EXPECT_CALL(*async_client, startRaw(_, _, _, _))...
         Expected: to be called once
           Actual: never called - unsatisfied and active

```

i guess some change in config 

https://dev.azure.com/cncf/envoy/_build/results?buildId=111237&view=logs&j=8c169225-0ae8-53bd-947f-07cb81846cb5&t=6b0ace90-28b2-51d9-2951-b1fd35e67dec&l=1233

**Comment by moderation** [comment]
@AlexanderEllis fantastic to see this land. I've been doing some testing sending from a bunch of Envoy's to Uptrace. On the OpenTelemetry Collector I have debug logging installed and it doesn't look like the traces are carrying any information. Similarly on the Uptrace side the spans are empty. 
```
Resource SchemaURL: 
ScopeSpans #0
ScopeSpans SchemaURL: 
InstrumentationScope  
Span #0
    Trace ID       : bec0620d406d5a2c63bede10ee2e29d6
    Parent ID      : 0cc2228c878e7a98
    ID             : d29c47da16a3c7f4
    Name           : router sync_service egress
    Kind           : SPAN_KIND_CLIENT
    Start time     : 2022-06-24 16:18:05.707695875 +0000 UTC
    End time       : 2022-06-24 16:18:06.370910412 +0000 UTC
    Status code    : STATUS_CODE_UNSET
    Status message : 
Span #1
    Trace ID       : bec0620d406d5a2c63bede10ee2e29d6
    Parent ID      : 
    ID             : 0cc2228c878e7a98
    Name           : egress xxxxx.xxxx.xx:9006
    Kind           : SPAN_KIND_CLIENT
    Start time     : 2022-06-24 16:18:05.707584509 +0000 UTC
    End time       : 2022-06-24 16:18:06.370936071 +0000 UTC
    Status code    : STATUS_CODE_UNSET
    Status message : 
```
In contrast my old Lightstep setup carried a bunch of info
image

Any idea on what is happening? Am I missing a config?

**Comment by AlexanderEllis** [comment]
Nothing wrong on your end! I'll be following this up with a few more PRs to flesh out the tracer, one of which will be the attribute setting correctly in the span (stubbed out span.setTag for now, but adding that [should allow these attributes to be set correctly in http_tracer_impl.cc](https://github.com/envoyproxy/envoy/blob/b67061197e30de9e9953abdff98dbcbbeea1f60e/source/common/tracing/http_tracer_impl.cc#L143)). I'll also add an example setup with the collector and a few docs as well.

Should hopefully be smooth(er) sailing now that the main PR is merged :)

**Comment by nareddyt** [comment]
https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/trace/v3/opentelemetry.proto mentions the following

> This extension is work-in-progress. Functionality is incomplete and it is not intended for production use. This extension has an unknown security posture and should only be used in deployments where both the downstream and upstream are trusted.

Are there any tangible steps to get this extension ready for production use? Is it a matter of adhering to Envoy's extension guidelines, or does it also involve the OpenTelemetry libraries?

https://github.com/envoyproxy/envoy/blob/main/EXTENSION_POLICY.md#extension-stability-and-security-posture

**Comment by htuch** [comment]
@itamarkam @yanavlasov can you address https://github.com/envoyproxy/envoy/pull/20281#issuecomment-1352329843? Thanks.

File: source/extensions/tracers/opentelemetry/tracer.cc
Comment (WojciechKuk) [inline source/extensions/tracers/opentelemetry/tracer.cc:115]: Is it possible to make it "less random"? i.e. 
hex timestamp (32b) + hex ms (16b) + hex instance's IP (32b) + some internal counter?

**Comment by itamarkam** [comment]
> itamark

@AlexanderEllis Any thoughts here?

**Pull request merged**
